File: examples/canonical-auth/elixir/lib/tenantcore_auth/auth/token_service.ex
defmodule TenantcoreAuth.Auth.TokenService do
  use Joken.Config

  alias TenantcoreAuth.Domain.InMemoryStore

  @issuer "tenantcore-auth"
  @audience "tenantcore-api"
  @expiry_seconds 900

  @impl true
  def token_config do
    default_claims(skip: [:exp, :iat, :nbf, :jti, :iss, :aud])
  end

  def issue_token(user, _store \\ InMemoryStore) do
    membership = InMemoryStore.get_active_membership(user.id)
    tenant_id = membership && membership.tenant_id
    tenant_role = membership && membership.tenant_role
    permissions =
      if tenant_role == "super_admin" do
        InMemoryStore.list_permissions()
        |> Enum.map(& &1.name)
        |> Enum.filter(&String.starts_with?(&1, "admin:"))
        |> Enum.sort()
      else
        InMemoryStore.get_effective_permissions(user.id)
      end
    groups =
      if tenant_id do
        InMemoryStore.get_groups_for_user(user.id, tenant_id) |> Enum.map(& &1.slug)
      else
        []
      end

    now = DateTime.utc_now() |> DateTime.to_unix()

    claims = %{
      "iss" => @issuer,
      "aud" => @audience,
      "sub" => user.id,
      "exp" => now + @expiry_seconds,
      "iat" => now,
      "nbf" => now,
      "jti" => Integer.to_string(System.unique_integer([:positive])),
      "tenant_id" => tenant_id,
      "tenant_role" => tenant_role,
      "groups" => groups,
      "permissions" => permissions,
      "acl_ver" => user.acl_ver
    }

    {:ok, token, _claims} = generate_and_sign(claims, signer())
    token
  end

  def sign_claims!(claims) do
    {:ok, token, _claims} = generate_and_sign(claims, signer())
    token
  end

  def verify_token(token) do
    with {:ok, claims} <- verify_and_validate(token, signer()),
         :ok <- verify_standard_claims(claims) do
      {:ok, claims}
    else
      _ -> {:error, :invalid_token}
    end
  end

  defp verify_standard_claims(claims) do
    now = DateTime.utc_now() |> DateTime.to_unix()
    exp = claims["exp"] || 0
    nbf = claims["nbf"] || 0

    cond do
      claims["iss"] != @issuer -> {:error, :issuer}
      claims["aud"] != @audience -> {:error, :audience}
      Enum.any?(~w[sub exp iat nbf jti tenant_role permissions acl_ver], &is_nil(claims[&1])) -> {:error, :claims}
      exp <= now -> {:error, :expired}
      nbf > now -> {:error, :not_before}
      true -> :ok
    end
  end

  defp signer do
    case System.get_env("TENANTCORE_TEST_SECRET") do
      nil -> Joken.Signer.create("HS256", "tenantcore-dev-secret")
      secret -> Joken.Signer.create("HS256", secret)
    end
  end
end

File: examples/canonical-auth/ruby/lib/tenantcore_auth/auth/token_service.rb
require "jwt"
require "openssl"
require "securerandom"

module TenantcoreAuth
  module Auth
    class TokenService
      ISSUER = "tenantcore-auth".freeze
      AUDIENCE = "tenantcore-api".freeze
      EXPIRY_SECONDS = 900

      def initialize(signing_key: nil)
        @signing_key = signing_key || self.class.signing_key
      end

      def issue_token(user, store)
        membership = store.get_active_membership(user.id)
        raise ArgumentError, "No active membership for #{user.id}" unless membership

        tenant_role = membership.tenant_role
        tenant_id = membership.tenant_id
        permissions =
          if tenant_role == "super_admin"
            store.admin_permissions
          else
            store.get_effective_permissions(user.id)
          end
        groups =
          if tenant_id
            store.get_groups_for_user(user.id, tenant_id).map(&:slug)
          else
            []
          end
        now = Time.now.to_i

        payload = {
          "iss" => ISSUER,
          "aud" => AUDIENCE,
          "sub" => user.id,
          "exp" => now + EXPIRY_SECONDS,
          "iat" => now,
          "nbf" => now,
          "jti" => SecureRandom.uuid,
          "tenant_id" => tenant_id,
          "tenant_role" => tenant_role,
          "groups" => groups,
          "permissions" => permissions,
          "acl_ver" => user.acl_ver
        }

        JWT.encode(payload, @signing_key, self.class.algorithm)
      end

      def verify_token(token)
        payload, _header = JWT.decode(
          token,
          self.class.verification_key,
          true,
          algorithm: self.class.algorithm,
          algorithms: [self.class.algorithm],
          iss: ISSUER,
          verify_iss: true,
          aud: AUDIENCE,
          verify_aud: true,
          verify_iat: true,
          verify_jti: true,
          verify_not_before: true,
          required_claims: %w[iss aud sub exp iat nbf jti tenant_role permissions acl_ver]
        )
        payload
      end

      def self.algorithm
        ENV["TENANTCORE_TEST_SECRET"] ? "HS256" : "RS256"
      end

      def self.signing_key
        return ENV["TENANTCORE_TEST_SECRET"] if ENV["TENANTCORE_TEST_SECRET"]

        @signing_key ||= OpenSSL::PKey::RSA.generate(2048)
      end

      def self.verification_key
        return ENV["TENANTCORE_TEST_SECRET"] if ENV["TENANTCORE_TEST_SECRET"]

        signing_key.public_key
      end
    end
  end
end
